### ISSUE TITLE
Resume a simulation

### ISSUE THREAD

**Issue by yiannisgaldadas** [openmm/openmm#3071]
Dear OpenMM community, 

I'm quite new to the OpenMM world and I'm currently trying to run a simple unbiased MD on a GPU cluster. I need, therefore, to be able to run the simulation in batches as there are time restrictions in the cluster I'm using. I have two questions:

1. Is there a way to append the StateDataReporters output of the restart to those of the previous step? So far, every time I run the next bit of the simulation, the .csv file starts from the beginning overwriting the previous .csv unless I change the name of the new .csv file.

2. OpenMM is very picky about the OpenMM version it was created with and even the type of GPU it was created on. It actually gives me version incompatibility errors when I try to restart the simulation, even though I'm running on the exact same computer with the exact same OpenMM version but probably different allocated GPUs as this depends on the cluster's availability at the time. I saw that restarting from a state.xml file could be a solution to this problem, but I couldn't find a reporter equivalent to the CheckpointReporter to create and .xml every few steps. Adding a simulation.saveState('state.xml') after the simulation.step(num_steps) doesn't help as the job is killed due to time limit before the simulation.saveState() is executed. 

Any help would be much appreciated as I'm a bit stuck with this. I have included below the piece of code I'm using at the moment to restart (unsuccessfully) a simulation.

Best,
Yiannis

`
# Specify the reporters parameters
num_steps = 500000000 # 1 us
trj_freq = 500
data_freq = 500
chk_freq = 5000

# Check whether to load a restart
restart = False
steps_so_far = 0
if os.path.isfile('checkpnt.chk'):
    restart = True
    print('Loading a restart')
    simulation.loadCheckpoint('checkpnt.chk')
    #simulation.loadState('checkpnt.xml')
    # find out how many steps to run left
    sim_log = [ i for i in open('data.csv').readlines() ]
    last_line = sim_log[-1]
    steps_so_far = last_line.split(',')[1] # how many steps run so far
    num_steps -= int(steps_so_far)

# if its a restart, will append new trj to the old one
simulation.reporters.append(app.DCDReporter('trajectory.dcd', trj_freq, append=restart))
simulation.reporters.append(app.StateDataReporter('data.csv', data_freq, step=True,
    potentialEnergy=True, temperature=True, volume=True, density=True,
    progress=True, remainingTime=True, speed=True, totalSteps=num_steps+int(steps_so_far)))
simulation.reporters.append(app.CheckpointReporter('checkpnt.chk', chk_freq))

# Run the MD simulation
print('Running Production...')
simulation.step(num_steps)`

**Comment by peastman** [comment]
> Is there a way to append the StateDataReporters output of the restart to those of the previous step?

The `file` argument passed to the StateDataReporter constructor can be either a filename or, alternatively, an open file object.  If you open the file in append mode, it should append:

```python
datafile = open('data.csv', 'a')
simulation.reporters.append(app.StateDataReporter(datafile, ...))
```

We should consider adding an `append` argument to StateDataReporter to make it do this automatically, similar to the `append` argument for DCDReporter.

> I couldn't find a reporter equivalent to the CheckpointReporter to create and .xml every few steps.

This would also be a good feature to add.  We could just add a `writeState` argument to tell it whether to write a state or a checkpoint.  The code changes would be very minor.  It would just change whether it opens the file in binary mode, and whether it writes a checkpoint or a serialized state.

I'll mark this issue as a feature request.  If you feel inclined to implement the changes, a PR would be welcome!

**Comment by yiannisgaldadas** [comment]
> The `file` argument passed to the StateDataReporter constructor can be either a filename or, alternatively, an open file object. If you open the file in append mode, it should append:

Many thanks for the tip, it worked like a charm!

> This would also be a good feature to add. We could just add a `writeState` argument to tell it whether to write a state or a checkpoint. The code changes would be very minor. It would just change whether it opens the file in binary mode, and whether it writes a checkpoint or a serialized state.

That would be pretty useful actually but I doubt my coding skills are good enough to do such a PR :P

**Comment by peastman** [comment]
Ok, I'll see if I can make these changes soon.

**Comment by yiannisgaldadas** [comment]
Apologies for bothering you again, but I still struggle with resuming my simulations. Even though the next step of the simulation is now appended to the original data.csv, the "steps" column starts from the beginning. Is this how it's supposed to be in OpenMM? I'm used to the Gromacs implementation, where the last step of the simulation is saved in the checkpoint and the total simulation time is stored in the .tpr so when you resume the simulation it continues from where it stopped. But in OpenMM, my understanding is that I need to calculate beforehand how many steps have already been run and how many more I need to run. 

```python
# Specify the reporters parameters
num_steps = 500000000 # 1 us
trj_freq = 500
data_freq = 500
chk_freq = 5000

# Check whether to load a restart
restart = False
steps_so_far = 0
if os.path.isfile('checkpnt.chk'):
    restart = True
    print('Loading a restart')
    #simulation.loadCheckpoint('checkpnt.chk')
    simulation.loadState('checkpnt.chk')
    # find out how many steps to run left
    sim_log = [ i for i in open('data.csv').readlines() ]
    last_line = sim_log[-1]
    steps_so_far = last_line.split(',')[1] # how many steps run so far
    num_steps -= int(steps_so_far)

# if its a restart, will append new trj to the old one
simulation.reporters.append(app.DCDReporter('trajectory.dcd', trj_freq, append=restart))
datafile = open('data.csv', 'a')
simulation.reporters.append(app.StateDataReporter(datafile, data_freq, step=True,
    potentialEnergy=True, temperature=True, volume=True, density=True,
    progress=True, remainingTime=True, speed=True, totalSteps=num_steps)))
simulation.reporters.append(app.CheckpointReporter('checkpnt.chk', chk_freq))

# Run the MD simulation
print('Running Production...')
simulation.step(num_steps)
```

**Comment by peastman** [comment]
This is implemented in #3249 and #3252.

I notice I never replied to your last question about step numbering.  Sorry about that!  In the current release, checkpoints and states don't record the step index.  But I recently changed that in #3248, so in the next release it will work the way you expect.

**Comment by yiannisgaldadas** [comment]
Many thanks for that! :)
 I'll keep an eye for the next release
